### PR TITLE
Add `Type::isConstantArray()`

### PR DESCRIPTION
**Pull request by herndlm**
Does not get rid of many `instanceof` type checks yet unfortunately though. in order to achieve that more methods on type are needed. I hope I can do a bit more soon, but there are definitely limits to what makes sense right now.

### CODE REVIEW

**Comment by herndlm** [comment]
failures seem to be unrelated packagist timeouts, I'll leave this alone for now..

File: src/Type/MixedType.php [hunk 1]
```diff
@@ -470,6 +470,17 @@ public function isArray(): TrinaryLogic
 		return TrinaryLogic::createMaybe();
 	}
 
+	public function isConstantArray(): TrinaryLogic
+	{
+		if ($this->subtractedType !== null) {
+			if ($this->subtractedType->isSuperTypeOf(new ConstantArrayType([], []))->yes()) {
+				return TrinaryLogic::createNo();
+			}
+		}
```
Comment (staabm) [inline src/Type/MixedType.php]: supporteded substraction is a new feature which should be covered by a test

Comment (herndlm) [reply src/Type/MixedType.php]: thx, I had a bad feeling about this and should have been more careful. trying to add tests made me realize that I have no clue how to implement this correctly for constant arrays, so I got rid of it and it returns always maybe now

Comment (staabm) [reply src/Type/MixedType.php]: hmm just guesswork, but maybe something like

```php
function doFoo($mixed) {
  $arr = ['a' => 1];
  if ($mixed != $arr) {
     assertType(...);
  }
}
```

or do you mean you don't know how the implementation should look like? :)

Comment (herndlm) [reply src/Type/MixedType.php]: more the latter. I think my initial implementation with the empty constant array is not correct

File: src/Type/MixedType.php [hunk 2]
```diff
@@ -470,6 +470,11 @@ public function isArray(): TrinaryLogic
 		return TrinaryLogic::createMaybe();
 	}
 
+	public function isConstantArray(): TrinaryLogic
+	{
+		return TrinaryLogic::createMaybe();
```
Comment (ondrejmirtes) [inline src/Type/MixedType.php]: Should say no if `isArray()` says no (it's subtracted).

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/MixedType.php (hunk 2 above)
Comment (herndlm) [reply src/Type/MixedType.php]: pushed a commit. just like that? I did not fully understand array subtraction here

Comment (ondrejmirtes) [reply src/Type/MixedType.php]: Yeah, make sense 😊

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**